File: app/src/main/java/it/cammino/risuscito/database/dao/CronologiaDao.java
package it.cammino.risuscito.database.dao;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import java.util.List;

import it.cammino.risuscito.database.CantoCronologia;
import it.cammino.risuscito.database.entities.Cronologia;

import static android.arch.persistence.room.OnConflictStrategy.REPLACE;

@Dao
public interface CronologiaDao {

  @Query("DELETE FROM cronologia")
  void truncateTable();

  @Query(
      "SELECT A.*, B.ultimaVisita FROM canto A, cronologia B WHERE A.id = B.idCanto ORDER BY B.ultimaVisita DESC")
  LiveData<List<CantoCronologia>> getLiveCronologia();

  @Query(
          "SELECT A.*, B.ultimaVisita FROM canto A, cronologia B WHERE A.id = B.idCanto ORDER BY B.ultimaVisita DESC")
  List<CantoCronologia> getCronologia();

  @Query("DELETE FROM cronologia")
  void emptyCronologia();

  @Update
  void updateCronologia(Cronologia cronologia);

  @Insert(onConflict = REPLACE)
  void insertCronologia(Cronologia cronologia);

  @Delete
  void deleteCronologia(Cronologia cronologia);

  @Delete
  void deleteCronologia(List<Cronologia> cronologiaList);
}

File: app/src/main/java/it/cammino/risuscito/database/dao/CronologiaDao.kt
package it.cammino.risuscito.database.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import it.cammino.risuscito.database.entities.Cronologia
import it.cammino.risuscito.database.pojo.CantoCronologia

@Suppress("unused")
@Dao
interface CronologiaDao {

    @Query("SELECT A.titolo, A.pagina, A.source, A.color, A.id, B.ultimaVisita FROM canto A, cronologia B WHERE A.id = B.idCanto ORDER BY B.ultimaVisita DESC")
    fun liveCronologia(): LiveData<List<CantoCronologia>>

    @Query("SELECT A.titolo, A.pagina, A.source, A.color, A.id, B.ultimaVisita FROM canto A, cronologia B WHERE A.id = B.idCanto ORDER BY B.ultimaVisita DESC")
    fun cronologia(): List<CantoCronologia>

    @Query("SELECT * FROM cronologia")
    fun all(): List<Cronologia>

    @Query("DELETE FROM cronologia")
    fun truncateTable()

    @Query("DELETE FROM cronologia")
    fun emptyCronologia()

    @Update
    fun updateCronologia(cronologia: Cronologia)

    @Upsert
    fun insertCronologia(cronologia: Cronologia)

    @Delete
    fun deleteCronologia(cronologia: Cronologia)

    @Delete
    fun deleteCronologia(cronologiaList: List<Cronologia>)
}
